java/ql/lib/semmle/code/java/ControlFlowGraph.qll

```diff
@@ -867,24 +939,50 @@ private module ControlFlowGraphImpl {
       // any other abnormal completion is propagated
       last(switch.getAStmt(), last, completion) and
       not completion instanceof YieldCompletion and
-      completion != NormalCompletion()
+      not completion instanceof NormalOrBooleanCompletion
     )
     or
-    // the last node in a case rule is the last node in the right-hand side
-    // if the rhs is a statement we wrap the completion as a break
-    exists(Completion caseCompletion |
-      last(n.(SwitchCase).getRuleStatement(), last, caseCompletion) and
+    // the last node in a case rule in statement context is the last node in the right-hand side.
+    // If the rhs is a statement, we wrap the completion as a break.
+    exists(Completion caseCompletion, SwitchStmt parent, SwitchCase case |
+      case = n and
+      case = parent.getACase() and
+      last(case.getRuleStatementOrExpressionStatement(), last, caseCompletion) and
       if caseCompletion instanceof NormalOrBooleanCompletion
       then completion = anonymousBreakCompletion()
       else completion = caseCompletion
     )
     or
-    // ...and if the rhs is an expression we wrap the completion as a yield
-    exists(Completion caseCompletion |
-      last(n.(SwitchCase).getRuleExpression(), last, caseCompletion) and
-      if caseCompletion instanceof NormalOrBooleanCompletion
-      then completion = YieldCompletion(caseCompletion)
-      else completion = caseCompletion
+    // ...and when a switch occurs in expression context, we wrap the RHS in a yield statement.
+    // Note the wrapping can only occur in the expression case, because a statement would need
+    // to have explicit `yield` statements.
+    exists(SwitchExpr parent, SwitchCase case |
+      case = n and
+      case = parent.getACase() and
+      (
+        exists(Completion caseCompletion |
+          last(case.getRuleExpression(), last, caseCompletion) and
+          if caseCompletion instanceof NormalOrBooleanCompletion
+          then completion = YieldCompletion(caseCompletion)
+          else completion = caseCompletion
+        )
+        or
+        last(case.getRuleStatement(), last, completion)
+      )
+    )
+    or
+    // The normal last node in a non-rule pattern case is the last of its variable declaration(s),
+    // or the successful matching of its guard if it has one.
+    // Note that either rule or non-rule pattern cases can end with pattern match failure, whereupon
+    // they branch to the next candidate pattern. This is accounted for in the `succ` relation.
+    exists(PatternCase pc | n = pc |
+      (
+        if exists(pc.getGuard())
+        then last(pc.getGuard(), last, BooleanCompletion(true, _))
+        else last(pc.getPattern(), last, NormalCompletion())
+      ) and
+      not pc.isRule() and
+      completion = NormalCompletion()
```

Comment:
I'd like to rework this, since the CFG design intends a clear split of responsibilities: `first` deals with half-edges entering an AST node, `last` deals with half-edges leaving an AST-node, and `succ` deals with those edges where both endpoints are within a given AST node. Thus enabling a nice structural recursion. The comment indicates that this split of responsibilities is broken here.
```suggestion
    exists(PatternCase pc | n = pc |
      last = n and completion = basicBooleanCompletion(false)
      or
      last(pc.getPattern(), last, NormalCompletion()) and
      not pc.isRule() and
      not exists(pc.getGuard()) and
      completion = basicBooleanCompletion(true)
      or
      last(pc.getGuard(), last, completion) and
      (
        completion = BooleanCompletion(true, _) and
        not pc.isRule()
        or
        completion = BooleanCompletion(false, _)
      )
```

Reply:
The suggested change should cover the half-edges we need in the `last` relation. It probably need a sprinkle of comments.